### PR TITLE
Fix the types (part 1)

### PR DESCRIPTION
**Pull request by rowanmanning**
Sorry for the size of this PR, it was really difficult to split but we're now at the point where we can introduce manual type definitions on a per-module basis. This will allow us to fix specific type issues that people are having when they come up.

I've added some tests to verify that we have a problem. I've tried to replicate a few different module formats that we use at the FT to make sure we can get this working with all of them:

  * In a CommonJS app (non-compiled)
  * In an [ESM](https://nodejs.org/api/esm.html) app (non-compiled)
  * In a TypeScript app that outputs CommonJS ([`esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) set to `false`)
  * In a TypeScript app that outputs ESM ([`esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) set to `true`)
  * In a TypeScript app that outputs CommonJS ([`esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) set to `false`)
  * In a TypeScript app that outputs ESM ([`esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) set to `true`)
  * In a CommonJS app which uses [Sucrase](https://sucrase.io/)

This is the first step towards having more confidence in our support of different module systems. This forms part of #303.

Various module formats don't work, e.g. the logger types don't work well for TypeScript projects that output CommonJS. The serialize-request types don't work for most ESM. I've added `@ts-ignore` statements to make the tests pass and document the places where we have inconsistency so that we can start to address specific type issues.

The last part of this PR is taking a currently working set of types (app-info) and reworking them so that the type definitions are manually created and exported. In other modules this will reduce the amount of JSDoc type declarations which is great. It also gives us better hinting in VSCode because we're naming the modules.

---

Relates-to: CPREL-972

### CODE REVIEW

File: packages/app-info/package.json [hunk 1]
```diff
@@ -14,5 +14,6 @@
     "node": "18.x || 20.x",
     "npm": "8.x || 9.x || 10.x"
   },
-  "main": "lib"
+  "main": "lib/index.js",
+  "types": "types/index.d.ts"
```
Comment (alexmuller) [inline packages/app-info/package.json:18]: Should this change also go in `scripts/create-package.js`?

Comment (rowanmanning) [reply packages/app-info/package.json:18]: Yep 👍

**Review by CyntiBinti — APPROVED** [review]
Praise: Definitely not an easy PR to have written especially going through the journey of getting it to this point - nice job on persevering with it 🙌🏽

**Review by alexmuller — APPROVED** [review]
Thanks!

**Pull request merged**